sylt-parser/src/parser.rs

```diff
@@ -6,13 +6,22 @@ use std::path::{Path, PathBuf};
 use sylt_common::error::Error;
 use sylt_common::rc::Rc;
 use sylt_common::Type as RuntimeType;
-use sylt_tokenizer::{file_to_tokens, Token};
+use sylt_tokenizer::{PlacedToken, Token, file_to_tokens};
 
 pub mod expression;
 pub mod statement;
 pub use self::expression::{Expression, ExpressionKind};
 pub use self::statement::{Statement, StatementKind};
 
+pub use sylt_tokenizer::Span;
+
+static INVALID_TOKEN: Token = Token::EOF;
+static INVALID_SPAN: Span = Span {
+    line: 0,
+    col_start: 0,
+    col_end: 0,
```

Comment:
Do we use only lines somewhere? Because this way to structure the information feels inferior to byte-offsets. Isn't it limiting to only have errors on one line - like if-statements and the like?

Reply:
The advantage of parsing the byte offsets into lines and columns early is that this way, only the tokenizer cares about byte offsets. If we only stored the byte offsets, the error reporter would have to calculate the line and column _for each singular error_, which feels wasteful since it doesn't care at all about the byte offsets.

Storing multiple lines can be done with two fields `line_start` and `line_end`. My idea was that spans could be joined, something like this:
```diff
         [T::If, ..] => {
+            let if_span = ctx.span();
+            // if_span = Span {
+            //     line_start: 2
+            //     line_end: 2
+            //     col_start: 4
+            //     col_end: 6
+            // }
             let (ctx, condition) = expression(ctx.skip(1))?;
  
             let (ctx, pass) = statement(ctx)?;
             // else?
             let (ctx, fail) = if matches!(ctx.token(), T::Else) {
                 let (ctx, fail) = statement(ctx.skip(1))?;
                 (ctx, fail)
             } else {
                 // No else so we insert an empty statement instead.
                 (
                     ctx,
                     Statement {
                         span: ctx.span(),
                         kind: EmptyStatement,
                     },
                 )
             };
  
+            let span = if_span.join(ctx.span());
+            // span = Span {
+            //     line_start: 2
+            //     line_end: 4
+            //     col_start: 4
+            //     col_end: 5
+            // }
             (
                 ctx,
                 If {
+                    span,
                     condition,
                     pass: Box::new(pass),
                     fail: Box::new(fail),
                 },
             )
         }

```

I didn't include this here since this felt self-contained.